csrc/multidevice/executor.cpp

```diff
@@ -6,133 +6,234 @@
  */
 // clang-format on
 #ifdef NVFUSER_DISTRIBUTED
+#include <ATen/cuda/CUDAContext.h>
+#include <device_lower/utils.h>
+#include <fusion_segmenter.h>
 #include <ir/utils.h>
-#include <multidevice/allocator.h>
+#include <multidevice/device_mesh.h>
 #include <multidevice/executor.h>
 #include <multidevice/lower_communication.h>
-#include <multidevice/pipeline.h>
+#include <multidevice/utils.h>
 
 namespace nvfuser {
 
-bool PipelineExecutor::shouldRun(PipelineStage* stage) {
-  if (should_run_.find(stage) == should_run_.end()) {
-    should_run_.emplace(
-        stage, stage->descriptor()->mesh.has(runtime_.comm_.deviceId()));
+namespace {
+
+std::pair<std::unique_ptr<Fusion>, std::unordered_map<Val*, Val*>>
```

Comment:
Please add a comment. I don't understand what this function is for.

Reply:
let me know if it is clearer this way. I also added a "TODO" comment for `allocateRecvBuffers`

Reply:
OK, I see. I did understand what it does, but I didn't why.